File: packages/shared-types/src/inbox.ts
// Inbox types — customer-centric planning inbox

import type { ScoreBreakdownItem } from './scoring';

export type LifecycleState = 'untouched' | 'active' | 'needs_action' | 'abandoned' | 'deleted';

export type InboxSortMode = 'rank_first' | 'due_date';

/**
 * A single customer row in the planning inbox.
 * Replaces the revision-centric InboxCandidate / CallQueueItem shape.
 */
export interface InboxItem {
  id: string;
  name: string | null;
  phone: string | null;
  email: string | null;
  street: string | null;
  city: string | null;
  postalCode: string | null;
  lat: number | null;
  lng: number | null;
  geocodeStatus: string;
  customerCreatedAt: string;

  // Lifecycle
  lifecycleState: LifecycleState;
  lifecycleRank: number;  // 0=untouched, 1=overdue, 2=upcoming, 3=needs_action

  // Next action (null for untouched/needs_action)
  nextActionKind: string | null;
  nextActionLabelKey: string | null;
  nextActionLabelFallback: string | null;
  nextActionDue: string | null;
  nextActionNote: string | null;

  // Contact history
  totalCommunications: number;
  lastContactAt: string | null;

  // Revision scheduling status ('scheduled' or 'confirmed' if visit agreed, else null)
  revisionStatus: string | null;

  // ID of the latest scheduled/confirmed revision (for unschedule action)
  latestScheduledRevisionId: string | null;

  // Count of scheduled/confirmed revisions (for ambiguity guard in unschedule)
  scheduledRevisionCount: number;

  // Urgency scoring (Phase 4+; 0 when scoring disabled)
  urgencyScore: number;
  /** Per-factor breakdown for the score explanation UI */
  scoreBreakdown?: ScoreBreakdownItem[];

  // Legacy device info (Phase 2-4; removed in Phase 6)
  deviceId?: string | null;
  deviceName?: string | null;
  deviceType?: string | null;
}

export interface InboxRequest {
  limit?: number;
  offset?: number;
  sortMode?: InboxSortMode;
  selectedRuleSetId?: string | null;
  // Filters
  lifecycleFilter?: LifecycleState[];
  geocodedOnly?: boolean;
  areaFilter?: string;
  /**
   * Deep-link focus: ask the worker to include this customer in the result set
   * even if they fall outside the normal top-N slice, and pin them to index 0.
   */
  focusCustomerId?: string;
}

export interface InboxResponse {
  items: InboxItem[];
  total: number;
  overdueCount: number;
  dueSoonCount: number;
  /**
   * Set by the worker when a focusCustomerId was requested.
   * true  = the focused customer is present in items (at index 0).
   * false = the focused customer could not be included (invalid/inaccessible).
   * absent/undefined = no focus was requested (backward-compatible).
   */
  focusedCustomerIncluded?: boolean;
}

File: worker/src/types/inbox.rs
#![allow(dead_code)]
//! Inbox types — customer-centric planning inbox

use chrono::{DateTime, NaiveDate, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

use crate::types::scoring::ScoreBreakdownItem;

/// Request to query the customer inbox
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct InboxRequest {
    pub limit: Option<u32>,
    pub offset: Option<u32>,
    /// Ignored — sorting is fully driven by the scoring profile (Phase 4B).
    /// Kept for backward compatibility; clients should stop sending this field.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sort_mode: Option<String>,
    pub selected_rule_set_id: Option<Uuid>,
    pub geocoded_only: Option<bool>,
    pub area_filter: Option<String>,
}

/// A single customer row in the planning inbox (Phase 2 shape).
/// `score_breakdown` is NOT in the DB — it is populated in Rust after scoring.
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
#[serde(rename_all = "camelCase")]
pub struct InboxItem {
    pub id: Uuid,
    pub name: Option<String>,
    pub phone: Option<String>,
    pub email: Option<String>,
    pub street: Option<String>,
    pub city: Option<String>,
    pub postal_code: Option<String>,
    pub lat: Option<f64>,
    pub lng: Option<f64>,
    pub geocode_status: String,
    pub customer_created_at: DateTime<Utc>,

    // Lifecycle
    pub lifecycle_state: String,
    pub lifecycle_rank: i32,

    // Next action (null for untouched/needs_action)
    pub next_action_kind: Option<String>,
    pub next_action_label_key: Option<String>,
    pub next_action_label_fallback: Option<String>,
    pub next_action_due: Option<NaiveDate>,
    pub next_action_note: Option<String>,

    // Contact history
    pub total_communications: i64,
    pub last_contact_at: Option<DateTime<Utc>>,

    // Revision scheduling status ('scheduled' or 'confirmed' if agreed, else null)
    pub revision_status: Option<String>,

    // Urgency score (Phase 4+; 0 when scoring disabled)
    pub urgency_score: f64,

    // Legacy device info (Phase 2-4)
    #[sqlx(default)]
    pub device_id: Option<Uuid>,
    #[sqlx(default)]
    pub device_name: Option<String>,
    #[sqlx(default)]
    pub device_type: Option<String>,

    /// Per-factor breakdown for the client-side explanation UI.
    /// NOT stored in the DB — always populated in Rust after scoring.
    /// Serialised as camelCase JSON in the NATS response.
    #[sqlx(skip)]
    pub score_breakdown: Vec<ScoreBreakdownItem>,
}

/// Response from the inbox query
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InboxResponse {
    pub items: Vec<InboxItem>,
    pub total: i64,
    pub overdue_count: i64,
    pub due_soon_count: i64,
}
